roachtest: extend github issue poster to include ip node mapping

williamchoe3 opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/138356

#### Motivation
Quality of life improvement, previously to find this information, you would need to find the name of the cluster in`test.log`, then use that cluster name to find it's cluster creation log in `{artifacts_dir}/_runner-logs/cluster-create`.

#### Approach
Get's node ip info from roachprod's `cloud.Cluster`
~~Similar to https://github.com/cockroachdb/cockroach/pull/151850~~
~~In the test cleanup phase in `inspectArtifacts` added a new function `gatherNodeIpMapping` which is best effort to gather the cluster information in the log.~~
* ~~First we need to find the right log file given that there could be retries for cluster creation which creates additional log files. Given the naming convention, I opted to the just sort the log files that contained the cluster name.  I realized a bit later though that because lexicographical sorting on numerical strings, this approach only works if retry attempts is <10. I was trying to avoid unnecessary string parsing if not needed. Also I would assume cluster retry attempts >=10 would be rare / too high of a retry limit to be ever set.~~
After finding the correct file, use regex to find the string in the log. Then store it in `test_impl`, then pass to `issues.Post` and I added a new case for parsing out the IP table
* The regex is flexible to number of columns so we can change the table fields without having to modify the regex 

Adds Cluster Node IP information to github issue e.g. 
```
| Node | Private IP | Public IP | Machine Type |
| --- | --- | --- | --- |
| willchoe-1758834520-01-n1cpu4-0001 | 10.142.0.2 | 34.139.44.53 | n2-standard-4 |
```
* Opted for keep the VM name as is vs. replacing with something like `n1`. My thinking was that if people wanted to reference the vm in logs, it would be nice to keep the names consistent, but if folks want this as n1 / the vm name isn't helpful then happy to change to `n1`, etc.

Also added a new `node-ips.log` that will also contain this table

#### Notes
Saw "cluster-create" being used as a magic string so created a const for it `clusterCreateDir = "cluster-create"`

#### Verification
Added datadriven test 

Manual Verification
Verified the renderer in `issues` was formatting the new code block correctly in debug mode, holding off on generating more debug github issues, but can if folks want to see

williamchoe3 committed:
roachtest: extend github issue poster to include ip node mapping

Quality of life improvement, previously to find this information, you would need
to find the name of the cluster in test.log, then use that cluster name to find
it's cluster creation log in artifacts_dir/_runner-logs/cluster-create.  Now
this information will be available in the github issue.

Resolves #138356

Epic: None
Release note: None

DarrylWong reviewed (commented):
I'd also (selfishly) like it to save the output of this mapping as a log. My general flow for investigating roachtests is that github info is just for cursory glances but once I start doing a deep dive (e.g. need to start matching ips to things in `cockroach.log`), I generally just stick to TeamCity and wouldn't want to go back to github to find the mapping.

srosenberg commented on pkg/cmd/roachtest/testdata/github/error_with_ip_node_info:
````diff
@@ -0,0 +1,66 @@
+Cluster Node to Ip Mapping:
+
+```
+Name                                DNS                                                                Private IP    Public IP       Machine Type   CPU Arch  CPU Family  Provision Model  
````
We definitely don't need the DNS column, and the name column can be abbreviated; nodes are always canonically named, i.e., `n1`, `n2`, etc. Thus, the most important are the three columns as illustrated in [1].

[1] https://github.com/cockroachdb/cockroach/pull/153962#issuecomment-3325417501

williamchoe3 replied:
redid my approach, no more log parsing. reference the new description for the new table. I'm open to changing the node names btw, there's a bullet in the description with my thoughts on it

DarrylWong commented:
I put out https://github.com/cockroachdb/cockroach/pull/154164 so hopefully 🤞  you should be able to test your changes without creating actual github issues.

srosenberg commented on pkg/cmd/bazci/githubpost/issues/condense.go:
```diff
@@ -55,6 +55,13 @@ type FatalNodeRoachtest struct {
 
+// NodeToIpMappingRoachtest contains the node to ip mapping from a roachtest
+// cluster
+type NodeToIpMappingRoachtest struct {
```
Nit: the naming convention below uses `roachtest` as prefix vs. the above suffix.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/utils_test.go:
```diff
@@ -27,3 +28,57 @@ func TestCmdLogFileName(t *testing.T) {
+			out, err := ToMarkdownTable(tt.input)
+			if err != nil {
+				if tt.shouldSucceed {
+					t.Errorf("ToMarkdownTable() failed when expected to succeed: error = %v", err)
```
Might as well check for the expected err since the above would also output the last row index in case of (column number) mismatch.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1265,11 +1265,16 @@ func (r *testRunner) runTest(
 
 				output := fmt.Sprintf("%s\ntest artifacts and logs in: %s", failureMsg, t.ArtifactsDir())
-				params := getTestParameters(t, issueInfo.cluster, issueInfo.vmCreateOpts)
+				// githubMsg may contain additional info not needed in output
```
Nit: the "not needed" modifier is confusing; the two outputs are mutually disjoint. Perhaps renaming `output` to `stdoutMsg` would suffice to disambiguate?

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1733,56 +1738,94 @@ func (r *testRunner) teardownTest(
-			return err
-		}
-		if out == "" {
+			// even if we encounter an error continue on, this is best effort
```
The comment seems to contradict the code below.

srosenberg commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1733,56 +1738,94 @@ func (r *testRunner) teardownTest(
+		t.L().Printf("Attempting to gather ip node mapping")
+		ipNodeMapOut, err := gatherNodeIpMapping(t, c)
+		if err != nil {
+			if joinErr := errors.Join(inspectArtifactsErr, err); joinErr != nil {
```
`err != nil` implies that `errors.Join(inspectArtifactsErr, err) != nil`. The nested if is redundant.

williamchoe3 commented on pkg/cmd/roachtest/roachtestutil/utils.go:
```diff
@@ -336,3 +336,37 @@ func SimulateMultiRegionCluster(
+	for _, row := range data[1:] {
+		if len(data[0]) != len(row) {
+			return "", errors.Errorf("row %d has %d columns, expected %d",
+				len(data)-1, len(row), len(data[0]))
```
fixed 😅

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1736,56 +1745,90 @@ func (r *testRunner) teardownTest(
+	// without a regex change
+	table = append(table, []string{"Node", "Private IP", "Public IP", "Machine Type"})
+	for _, vmInstance := range cachedCluster.VMs {
+		table = append(table, []string{vmInstance.Name, vmInstance.PrivateIP, vmInstance.PublicIP, vmInstance.MachineType})
```
In hindsight, we probably don't need the machine type since its static, i.e. I can just look at the test to figure it out

srosenberg replied:
Agreed, machine type is easily inferred and not really relevant in majority of (triage) cases. Let's remove.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1736,56 +1745,90 @@ func (r *testRunner) teardownTest(
 	// Works with local and remote node clusters because we will always download
 	// the artifacts if there's a test failure (except for timeout)
-	cmd := exec.Command("grep", args...)
+	cmd := exec.Command(command, args...)
```
nit: change seems redundant since command is never modified, makes it marginally harder to follow

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1736,56 +1745,90 @@ func (r *testRunner) teardownTest(
 	filePattern := "logs/*unredacted/cockroach*.log"
-	// *unredacted captures patterns for single node and multi-node clusters
-	// e.g. unredacted, 1.unredacted
+	// * wildcard to capture patterns for single node and multi-node clusters
```
nit: not a big deal but some of these fatal node log fixes could've been their own commit

srosenberg commented on pkg/cmd/roachtest/github_test.go:
```diff
@@ -251,6 +249,18 @@ F250826 19:49:07.194443 3106 sql/sem/builtins/builtins.go:6063 ⋮ [T1,Vsystem,n
+				case "ip-node-info":
+					testCase.message = fmt.Sprintf("%s\n%s", testCase.message, `| Node | Private IP | Public IP | Machine Type |
+| --- | --- | --- | --- |
+| teamcity-1758834520-01-n1cpu4-0001 | 10.142.0.2 | 34.139.44.53 | n2-standard-4 |`)
```
One tiny nit... Could we swap the positions of public and private ips? In most infra flakes I can think of, a public rather than private ip is implicated. So, I worry about the case where the cluster name is very long and the public ip is obscured by the scrollbar; i.e., this optimizes for (clusterName, publicIp) to be always visible, even on tiny screens :)

williamchoe3 replied:
sure np

srosenberg commented on pkg/cmd/bazci/githubpost/issues/formatter_unit.go:
```diff
@@ -74,6 +74,17 @@ func (unitTestFormatterTyp) Body(r *Renderer, data TemplateData) error {
 		r.CodeBlock("", fnr.Message)
 		r.Escaped("Fatal entries found in Cockroach logs:")
 		r.CodeBlock("", fnr.FatalLogs)
+		if nodeIpMap, ok := data.CondensedMessage.NodeToIpMappingRoachtest(); ok {
```
I see it's effectively demultiplexing. But then we have a bug in `getGithubMessage` which overwrites instead of appends, in the case of fatal logs?

williamchoe3 replied:
```
	githubMsg := failureMsg
	if githubFatalLogs := t.getGithubFatalLogs(); githubFatalLogs != "" {
		githubMsg = fmt.Sprintf("%s\n%s", githubMsg, githubFatalLogs)
	}
	if githubIpToNodeMapping := t.getGithubIpToNodeMapping(); githubIpToNodeMapping != "" {
		githubMsg = fmt.Sprintf("%s\n%s", githubMsg, githubIpToNodeMapping)
	}
```

I'm doing a weird append, I think i did it this way because I needed the`\n` prepended to `githubMsg`, but also a `\n` char inbetween the two info types if they both existed

williamchoe3 commented:
Another Smoke Test after addressing comments + merge conflicts https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20707055?hideTestsFromDependencies=false&hideProblemsFromDependencies=false

srosenberg commented on pkg/cmd/roachtest/github_test.go:
```diff
@@ -228,6 +228,13 @@ func TestCreatePostRequest(t *testing.T) {
 test artifacts and logs in: artifacts/roachtest/manual/monitor/test-failure/node-fatal-explicit-monitor/cpu_arch=arm64/run_1
 F250826 19:49:07.194443 3106 sql/sem/builtins/builtins.go:6063 ⋮ [T1,Vsystem,n1,client=127.0.0.1:54552,hostssl,user=‹roachprod›] 250  force_log_fatal(): ‹oops›
+`)
+						case "error-with-ip-node-info":
```
You mean #157021, right?
